### PR TITLE
feat: Add documentation for UI Unit Testing with Spring

### PR DESCRIPTION
**Pull request by mcollovati**
Part of vaadin/testbench#1339

### CODE REVIEW

**Comment by mcollovati** [comment]
Follows #1413

File: articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+}
+----
+
+[[testbench.uiunit.setup-view-access-control]]
+== Setup View Access Control
+
+To apply view access control, Vaadin requires a [classname]`ViewAccessChecker` to be registered as [classname]`BeforeEnterListener` for the UI.
+For [annotationname]`@SpringBootTest` annotated tests, the checker is automatically created and configured.
+However, when testing with a restricted ApplicationContext, you may want to perform the setup by your own in a [classname]`Configuration` class,
+by providing a [classname]`VaadinServiceInitListener` that executes the step mentioned above.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc:51]:
```suggestion
== Set Up View Access Control

To apply view access control, Vaadin requires a [classname]`ViewAccessChecker` to be registered as a [classname]`BeforeEnterListener` for the UI.
For [annotationname]`@SpringBootTest` annotated tests, the checker is automatically created and configured.
However, when testing with a restricted ApplicationContext, you may want to perform the setup yourself in a [classname]`Configuration` class by providing a [classname]`VaadinServiceInitListener` that executes this step.
```

File: articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+
+.Setup ViewAccessChecker for plain Spring project
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc:54]:
```suggestion
.Set Up a ViewAccessChecker for a plain Spring project
```

File: articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        return event -> {
+            event.getSource().addUIInitListener(uiEvent -> {
+                uiEvent.getUI().addBeforeEnterListener(viewAccessChecker);
+            });
+        };
+    }
+}
+----
+
+If using Vaadin Spring Add-On, you can instead import the out-of-the box [classname]`ViewAccessCheckerInitializer`, that only requires to define a [classname]`ViewAccessChecker` bean.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc:73]:
```suggestion
If you are using the Vaadin Spring Add-On, you can instead import the out-of-the box [classname]`ViewAccessCheckerInitializer`, which only requires you to define a [classname]`ViewAccessChecker` bean. 
```

File: articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    }
+}
+----
+
+
+[[testbench.uiunit.spring-security]]
+== Testing With Spring Security
+
+Vaadin comes with built-in security helpers that enable annotation based view access control mechanism, which integrates well with Spring Security.
+When using [classname]`SpringUIUnitTest`, if Spring Security is present on classpath, the mock environment is instructed to fetch authentication details from Spring [classname]`SecurityContextHolder`.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc:94]:
```suggestion
Vaadin comes with built-in security helpers that enable the annotation-based view access control mechanism, which integrates well with Spring Security.
When using [classname]`SpringUIUnitTest`, if Spring Security is present on the classpath, the mock environment is instructed to fetch authentication details from Spring [classname]`SecurityContextHolder`.
```

File: articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+
+With this support, you can use Spring Security test annotations, such as [annotationname]`@WithMockUser`, [annotationname]`@WithAnonymousUser` or [annotationname]`@WithUserDetails`, to simulate different authentication scenarios with test method granularity.
+More information is available at https://docs.spring.io/spring-security/reference/servlet/test/method.html#test-method-withmockuser[Spring Security documentation] site.
+Authentication details are available before creating the UI instance and navigating to the default route, so redirects to login view does not happen, when simulating logged in users.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc:99]:
```suggestion
More information is available on the https://docs.spring.io/spring-security/reference/servlet/test/method.html#test-method-withmockuser[Spring Security documentation] site.
Authentication details are available before creating the UI instance and navigating to the default route, so redirects to the login view are not performed when simulating logged-in users.
```

File: articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+----
+<dependency>
+        <groupId>org.springframework.security</groupId>
+        <artifactId>spring-security-test</artifactId>
+        <scope>test</scope>
+</dependency>
+----
+
+Then extend [classname]`SpringUIUnitTest` and annotate test methods on order to setup an authentication scenario.
+For simplest use cases, use [annotationname]`@WithMockUser` or [annotationname]`@WithAnonymousUser`, providing username and roles that should be granted.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc:115]:
```suggestion
Then extend [classname]`SpringUIUnitTest` and annotate test methods in order to set up an authentication scenario.
For the simplest use cases, use [annotationname]`@WithMockUser` or [annotationname]`@WithAnonymousUser`, providing the username and roles that should be granted.
```

File: articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+.Tests with mocked users
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc:117]:
```suggestion
.Tests with mock users
```

File: articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        navigate(AdminRoleView.class);
+
+        Assertions.assertTrue(
+                $(Avatar.class).first().getComponent().isVisible(),
+                "Avatar should be visible for logged users");
+    }
+}
+----
+
+When custom User objects or complex grant rules should be used, provide a custom [classname]`UserDetailsService` and annotate test method with [annotationname]`@WithUserDetails`.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc:152]:
```suggestion
When custom User objects or complex grant rules should be used, provide a custom [classname]`UserDetailsService` and annotate the test method with [annotationname]`@WithUserDetails`.
```

File: articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+.Tests with mocked UserDetailsService
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing/spring-tests.asciidoc:154]:
```suggestion
.Tests with mock UserDetailsService
```

**Review by caalador — APPROVED** [review]

**Comment by tarekoraby** [comment]
@caalador @mcollovati for future reference, feel free to merge PRs if the language check has passed (if you are in a hurry, you can even merge before the language check is completed, and we can do the language editing in a separate PR if something needs editing).

**Pull request merged**